### PR TITLE
HPCC-32649 Avoid writing any data to an empty compressed file

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [x] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by ghalliday** [comment]
Push for regression testing.
NOTE: This cannot be merged until all systems that might read these files are upgraded to a compatible version.

File: system/jlib/jlzw.cpp [hunk 1]
```diff
@@ -2456,19 +2456,23 @@ class CCompressedFile : implements ICompressedFileIO, public CInterface
...
-                trailer.crc = crc32((const char *)indexbuf.toByteArray(),
-                                indexbuf.length(),trailer.crc);
-                indexbuf.append(trailer.crc);
-            }
-            else {
-                trailer.datacrc = 0;
-                trailer.crc = ~0U;
-                indexbuf.append(sizeof(trailer),&trailer);
+            //Avoid writing out a header/footer if the file is empty
+            if (indexbuf.length() != 0)
```
Comment (jakesmith) [inline system/jlib/jlzw.cpp:2460]: would testing trailer.expandedSize == 0 be better and mean can avoid call to flush() (on line 2458)

**Review by jakesmith — COMMENTED** [review]
@ghalliday - 1 question.

Also, perhaps should add a configurable option now (disabled by default), so we can merge the code, but enable it later?

File: system/jlib/jlzw.cpp (hunk 1 above)
Comment (ghalliday) [reply system/jlib/jlzw.cpp]: flush returns immediately, but I think you're right it would be clearer.

**Comment by ghalliday** [comment]
@jakesmith repushed and rebased.  
It now allows the previous functionality to be supported by setting an expert option, but defaulting on for 9.10.x

**Review by jakesmith — APPROVED** [review]
@ghalliday - looks good.

**Pull request merged**